### PR TITLE
Improve shuffle quality

### LINKED ISSUE #1256: thrust::shuffle gives low randomness permutations
The thrust::shuffle function gives low randomness permutations starting from small vector sizes (approximately from N = 150, the exact value depends on the seed), below are shown some outputs with this behavior it doesn't matter if you change the seed or the randomness engine.

Version tested: Last _main_ commit d83ea9

Example outputs varying N:
```
N = 10 // OK
5 4 7 6 1 0 3 2 9 8 

N = 100 // OK
78 94 46 62 14 30 79 95 47 63 15 31 76 92 44 60 12 28 77 93 45 61 13 29 74 90 42 58 10 26 75 91 43 59 11 27 72 88 40 56 8 24 73 89 41 57 9 25 70 86 38 54 6 22 71 87 39 55 7 23 68 84 36 52 4 20 69 85 37 53 5 21 98 66 82 34 50 2 18 99 67 83 35 51 3 19 96 64 80 32 48 0 16 97 65 81 33 49 1 17

N = 150 // low randomness
149 148 145 144 147 146 133 132 135 134 129 128 131 130 141 140 143 142 137 136 139 138 21 20 23 22 17 16 19 18 29 28 31 30 25 24 27 26 5 4 7 6 1 0 3 2 13 12 15 14 9 8 11 10 53 52 55 54 49 48 51 50 61 60 63 62 57 56 59 58 37 36 39 38 33 32 35 34 45 44 47 46 41 40 43 42 85 84 87 86 81 80 83 82 93 92 95 94 89 88 91 90 69 68 71 70 65 64 67 66 77 76 79 78 73 72 75 74 117 116 119 118 113 112 115 114 125 124 127 126 121 120 123 122 101 100 103 102 97 96 99 98 109 108 111 110 105 104 107 106

N = 200 // low randomness
149 148 151 150 145 144 147 146 157 156 159 158 153 152 155 154 133 132 135 134 129 128 131 130 141 140 143 142 137 136 139 138 181 180 183 182 177 176 179 178 189 188 191 190 185 184 187 186 165 164 167 166 161 160 163 162 173 172 175 174 169 168 171 170 197 196 199 198 193 192 195 194 21 20 23 22 17 16 19 18 29 28 31 30 25 24 27 26 5 4 7 6 1 0 3 2 13 12 15 14 9 8 11 10 53 52 55 54 49 48 51 50 61 60 63 62 57 56 59 58 37 36 39 38 33 32 35 34 45 44 47 46 41 40 43 42 85 84 87 86 81 80 83 82 93 92 95 94 89 88 91 90 69 68 71 70 65 64 67 66 77 76 79 78 73 72 75 74 117 116 119 118 113 112 115 114 125 124 127 126 121 120 123 122 101 100 103 102 97 96 99 98 109 108 111 110 105 104 107 106
```

Reproducer code:
```
#include <thrust/shuffle.h>

int main()
{
        const int N = 150;
        int result[N];

        int seed = 0;

        thrust::shuffle_copy(
                thrust::counting_iterator<int>(0),
                thrust::counting_iterator<int>(N),
                result,
                thrust::default_random_engine(seed));

        for(size_t I=0;I<N;I++)
                printf("%d ", result[I]);
        printf("\n");

        return 0;
}
```

### PR DESCRIPTION
**Pull request by RAMitchell**
This is another version of #1257, without using any proprietary hash functions.

I have included thrust::random::ranlux48 as an additional hash, as well as improved the mixing function (before I used XOR alone, which turns out to be a [poor choice](https://stackoverflow.com/a/27952689)) to match boost::hash_combine.

I also increased the number of rounds from 8->16.

This passes all tests from #1257 and fixes #1256

@djns99 please review.

### CODE REVIEW

File: thrust/system/detail/generic/shuffle.inl
Comment (djns99) [inline thrust/system/detail/generic/shuffle.inl:92]: I think this is intended for 32 bit numbers. Though this is probably fine since the inputs can most likely be uint32_t, but we would get less randomness from the key so its a trade-off.
I believe boost uses a variant murmur hash 2 for uint64_t.

```
        inline void hash_combine_impl(boost::uint64_t& h,
                boost::uint64_t k)
        {
            const boost::uint64_t m = UINT64_C(0xc6a4a7935bd1e995);
            const int r = 47;

            k *= m;
            k ^= k >> r;
            k *= m;

            h ^= k;
            h *= m;

            // Completely arbitrary number, to prevent 0's
            // from hashing to 0.
            h += 0xe6546b64;
        }
```

File: thrust/system/detail/generic/shuffle.inl
Comment (djns99) [inline thrust/system/detail/generic/shuffle.inl:121]: Should this be constexpr? Im not sure the standards but generally its a good idea if you can isnt it? That way the object is smaller since it doesnt need the field (not that it matters really).

File: thrust/system/detail/generic/shuffle.inl
Comment (djns99) [inline thrust/system/detail/generic/shuffle.inl]: I believe there is a bug here where powers of 2 will be rounded up again.
We need to keep the guard for m == 0 and m--

File: thrust/system/detail/generic/shuffle.inl
Comment (RAMitchell) [reply thrust/system/detail/generic/shuffle.inl:92]: Ok will use this one.

File: thrust/system/detail/generic/shuffle.inl
Comment (RAMitchell) [reply thrust/system/detail/generic/shuffle.inl:121]: I'll make it a template parameter again.

File: thrust/system/detail/generic/shuffle.inl
Comment (RAMitchell) [reply thrust/system/detail/generic/shuffle.inl]: Good catch, any other fixes in your PR that I need to include?

Comment (djns99) [reply thrust/system/detail/generic/shuffle.inl]: Not that I could see. I looked at the commit diff for the review and didn't notice anything

File: thrust/system/detail/generic/shuffle.inl
Comment (RAMitchell) [reply thrust/system/detail/generic/shuffle.inl:94]: Using the above function causes tests to fail for some reason. I will keep my version for now.

Comment (djns99) [reply thrust/system/detail/generic/shuffle.inl:94]: If you are going to do that might be worth changing the key to uint32_t so it takes less space

Comment (RAMitchell) [reply thrust/system/detail/generic/shuffle.inl:94]: We can leave that for future work.

Comment (djns99) [reply thrust/system/detail/generic/shuffle.inl:94]: Ok no worries. I am happy with it then.

File: testing/shuffle.cu
Comment (allisonvacanti) [inline testing/shuffle.cu:11]: Can you add some comments here, explaining what this does and why it's needed? As a maintainer, it's not obvious what this is testing, and the implementation is rather cryptic on its own.

File: testing/shuffle.cu
Comment (allisonvacanti) [inline testing/shuffle.cu:323]: Unrelated, should be reverted from this PR.

File: testing/shuffle.cu
Comment (allisonvacanti) [inline testing/shuffle.cu:339]: Please revert. Thrust style prefers `template <...>` on a separate line.

File: testing/shuffle.cu
Comment (allisonvacanti) [inline testing/shuffle.cu:355]: Style.

File: testing/shuffle.cu
Comment (allisonvacanti) [inline testing/shuffle.cu:405]: Revert

Comment (allisonvacanti) [reply testing/shuffle.cu:405]: There are several more unrelated changes in this PR -- I'll stop pointing them out, but please revert them.

(I have a tentative plan to do a project-wide restyling that will preserve history, so I'd rather not touch lines that don't relate to the current PR topic.)

**Review by allisonvacanti — DISMISSED** [review]
Thanks for redoing this!

I've requested a few minor changes (mostly style and docs) inline, once those are addressed I'll start testing.

File: testing/shuffle.cu
Comment (djns99) [reply testing/shuffle.cu:405]: I'm just curious, is there a clang-format or similar file that describes the target style for future reference?
I don't remember seeing anything, though I didn't look too hard.

Comment (allisonvacanti) [reply testing/shuffle.cu:405]: Not yet. The current style of Thrust and CUB is pretty wild and changes from file to file, so it's sort of ill-defined. I have a clang-format file that I use locally, but it's not definitive (and unfortunately, clang-format has no idea what to do with things like `__host__` and `__device__`, so we currently can't rely on it).

New code coming in can be whatever style is reasonable and matches what's already in the file, but I want to hold off on restyling existing code for now to preserve history when we do switch to clang-format.

Comment (RAMitchell) [reply testing/shuffle.cu:405]: Good to know. Would be great for external contributors to have clang-format eventually.

Comment (allisonvacanti) [reply testing/shuffle.cu:405]: Agreed -- it'd be great for everyone, internal and external! :)

**Comment by allisonvacanti** [comment]
LGTM, thanks for the quick updates!

DVS CL 29194481.

**Comment by allisonvacanti** [comment]
This is failing our internal CI. Can you take a look at this error? Looks like an easy fix.

```

/dvs/p4/build/sw/gpgpu/thrust/testing/unittest/assertions.h: In instantiation of 'void unittest::assert_equal(T1, T2, const string&, int) [with T1 = long unsigned int; T2 = int; std::__cxx11::string = std::__cxx11::basic_string]':
/dvs/p4/build/sw/gpgpu/thrust/testing//shuffle.cu:407:109: required from here
/dvs/p4/build/sw/gpgpu/thrust/testing/unittest/assertions.h:120:9: error: comparison of integer expressions of different signedness: 'long unsigned int' and 'int' [-Werror=sign-compare]
```

**Comment by allisonvacanti** [comment]
CI found some janky casts in the cephes code. Added some alignment hints to make sure these work out in practice.

New DVS CL: 29278660

**Comment by allisonvacanti** [comment]
New DVS CL: 29283398

**Comment by allisonvacanti** [comment]
More XLC fixes, new DVS CL: 29287861

**Comment by allisonvacanti** [comment]
New DVS CL: 29301680

File: testing/shuffle.cu
Comment (allisonvacanti) [inline testing/shuffle.cu:141]: We need to come up with a different implementation for this. I'm not sure why this implementation is written this way, but it's defining a buffer of 16-bit ints that are then bitcast into 64-bit floats.

Comment (allisonvacanti) [reply testing/shuffle.cu:141]: I've tried using some alignment hints to work around this, but some compilers still aren't happy, and tbh making these sorts of assumptions about alignment, endianness, and type representations makes me nervous.

@RAMitchell @djns99 please find an alternate implementation for this -- if we can get something that passes tests by Friday I can include this PR in 1.11.0, otherwise it may slip to 1.12.0.

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

File: testing/shuffle.cu
Comment (djns99) [reply testing/shuffle.cu:141]: @RAMitchell @allisonvacanti 
I am not able to test right now, but here are the tables as a double. This should resolve the issues I believe:
```
static constexpr double A[] = {0.000811614167470508488140545910738410384510643780,-0.000595061904284301438315674115386855191900394857,0.000793650340457716942620114419781884862459264696,-0.002777777777300996942672073330982129846233874559,0.083333333333333189929525985917280195280909538269};
static constexpr double B[] = {-1378.251525691208598800585605204105377197265625,-38801.631513463784358464181423187255859375,-331612.9927388711948879063129425048828125,-1162370.97492762305773794651031494140625,-1721737.00820839661173522472381591796875,-853555.66424576542340219020843505859375};
static constexpr double C[] = {-351.8157014365234545039129443466663360595703125,-17064.21066518811494461260735988616943359375,-220528.59055385444662533700466156005859375,-1139334.44367982516996562480926513671875,-2532523.07177582941949367523193359375,-2018891.4143353276886045932769775390625};
```

**Comment by allisonvacanti** [comment]
Thanks for quick fix -- restarted testing under CL 29315896

**Comment by allisonvacanti** [comment]
Tests are passing 👍 Merging now.

Thanks to both of you for helping me get this PR over all of the hurdles :)

**Review by allisonvacanti — APPROVED** [review]

**Comment by allisonvacanti** [comment]
@RAMitchell @djns99 QA noticed that this patch drops the performance of `thrust::shuffle` by 60-95%. I don't think we want to revert the patch over this -- the shuffle quality improvements are worth paying a bit of perf, so I'll have them update their baselines.

Would either of you be interested in optimizing the new implementation to see if some of the performance can be recovered?

**Comment by RAMitchell** [comment]
Yep, we will be doing more work on it over summer. Most importantly we now have an accurate baseline. Can I ask how the 60-95% is measured?

**Comment by allisonvacanti** [comment]
They're running the shuffle benchmarks in `internal/benchmark/bench.cu` and comparing the average throughput before and after this PR.

**Pull request merged**
